content/advent-2015/test2doc-generate-your-api-docs.md

```diff
@@ -0,0 +1,294 @@
+	//      where the returned map is of the form map[key]value
+	test.RegisterURLVarExtractor(myURLVarExtractorFn)
+
+
+	// 2. You must use test2doc/test's wrapped httptest.Server instead of
+	//    the raw httptest.Server, so that test2doc can listen to and
+	//    record requests & responses.
+	//
+	//    NewServer takes your HTTP handler as an argument
+	server, err := test.NewServer(router)
+	if err != nil {
+		panic(err.Error())
+	}
+
+	// .. then run your tests as usual
+	exitCode := m.Run()
+
+
+	// 3. Finally, you must tell the wrapped server when you are done testing
+	//    so that the buffer can be flushed to an apib doc file
+	server.Finish()
+
+	// note that os.Exit does not respect defers.
+	os.Exit(exitCode)
+}
+
+```
+
+Some example `URLVarExtractor`s for different routers:
+
+`gorilla/mux`:
+
+```go
+// NOTE: if you are using gorilla/mux, you must set the 
+// router's 'KeepContext' to true, so that url parameters 
+// can be accessed after the request has been handled.
+router.KeepContext = true
+    
+// Use mux.Vars func as URLVarExtractor
+test.RegisterURLVarExtractor(mux.Vars)
+```
+<br>
+
+`julienschmidt/httprouter`:
+
+```go
+import (
+	"net/http"
+
+	"github.com/adams-sarah/test2doc/doc/parse"
+	"github.com/julienschmidt/httprouter"
+)
+
+// MakeURLVarExtractor returns a func which extracts 
+// url vars from a request for test2doc documentation generation
+func MakeURLVarExtractor(router *httprouter.Router) parse.URLVarExtractor {
+	return func(req *http.Request) map[string]string {
+		// httprouter Lookup func needs a trailing slash on path
+		path := req.URL.Path
+		if !strings.HasSuffix(path, "/") {
+			path += "/"
+		}
+
+		_, params, ok := router.Lookup(req.Method, path)
+		if !ok {
+			return nil
+		}
+
+		paramsMap := make(map[string]string, len(params))
+		for _, p := range params {
+			paramsMap[p.Key] = p.Value
+		}
+
+		return paramsMap
+	}
+}
+
+// and then..
+test.RegisterURLVarExtractor(MakeURLVarExtractor(router))
+```
+<br>
+
+## How the magic happens
+
+### Recording Requests and Responses:
+Recording the requests and responses was pretty straight-forward:
+
+1. Write the HTTP handler's response initially to an `http.ResponseRecorder`
+2. Add the response header/body to the documentation
+3. Copy the response back to the original `http.ResponseWriter`
+4. Continue executing the test
+
+
+### Fetching the handler's Go doc string:
+First, let's take a look at the `http.ResponseWriter` interface
+from Go's `net/http` package:
+
+```go
+type ResponseWriter interface {
+        Header() Header
+        Write([]byte) (int, error)
+        WriteHeader(int)
+}
+```
+
+
+test2doc has its own `ResponseWriter`, which implements
+`http.ResponseWriter`, and looks something like this:
+
+```go
+type ResponseWriter struct {
+	HandlerInfo HandlerInfo
+	URLVars     map[string]string
+	W           *httptest.ResponseRecorder
+}
+```
+
+The test2doc `ResponseWriter`'s implementations of the `Header` and 
```

Comment:
Updated to: 
"Test2doc's `ResponseWriter` implements the `Header` and 
`WriteHeader` methods by just falling back those of its 
embedded `httptest.ResponseRecorder`"